// mdb-wysiwyg/bundles/mdb-wysiwyg.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/forms'), require('@angular/common'), require('angular-bootstrap-md')) :
    typeof define === 'function' && define.amd ? define('mdb-wysiwyg', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/forms', '@angular/common', 'angular-bootstrap-md'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['mdb-wysiwyg'] = {}, global.ng.core, global.rxjs, global.rxjs.operators, global.ng.forms, global.ng.common, global['angular-bootstrap-md']));
}(this, (function (exports, core, rxjs, operators, forms, common, angularBootstrapMd) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TEXT_STYLE_ITEMS = [
        { type: 'paragraph', selector: 'p', text: 'Paragraph' },
        { type: 'heading', selector: 'h1', text: 'Heading' },
        { type: 'heading', selector: 'h2', text: 'Heading' },
        { type: 'heading', selector: 'h3', text: 'Heading' },
        { type: 'heading', selector: 'h4', text: 'Heading' },
        { type: 'heading', selector: 'h5', text: 'Heading' },
        { type: 'heading', selector: 'h6', text: 'Heading' },
        { type: 'preformatted', selector: 'pre', text: 'Preformatted' },
    ];
    var TEXT_COLOR_ITEMS = [
        { name: 'red', color: '#f44336' },
        { name: 'pink', color: '#e91e63' },
        { name: 'purple', color: '#9c27b0' },
        { name: 'deep-purple', color: '#673ab7' },
        { name: 'indigo', color: '#3f51b5' },
        { name: 'blue', color: '#2196f3' },
        { name: 'light-blue', color: '#03a9f4' },
        { name: 'cyan', color: '#00bcd4' },
        { name: 'teal', color: '#009688' },
        { name: 'green', color: '#4caf50' },
        { name: 'light-green', color: '#8bc34a' },
        { name: 'lime', color: '#cddc39' },
        { name: 'yellow', color: '#ffeb3b' },
        { name: 'amber', color: '#ffc107' },
        { name: 'orange', color: '#ff9800' },
        { name: 'deep-orange', color: '#ff5722' },
        { name: 'brown', color: '#795548' },
        { name: 'grey', color: '#9e9e9e' },
        { name: 'blue-grey', color: '#607d8b' },
        { name: 'mdb-color', color: '#45526e' },
        { name: 'white', color: '#fff' },
        { name: 'black', color: '#000' },
    ];
    var TEXT_DECORATION_ITEMS = [
        { type: 'bold', icon: 'fas fa-bold', tooltip: 'Bold' },
        { type: 'italic', icon: 'fas fa-italic', tooltip: 'Italic' },
        { type: 'strikethrough', icon: 'fas fa-strikethrough', tooltip: 'Strikethrough' },
        { type: 'underline', icon: 'fas fa-underline', tooltip: 'Underline' },
        { type: 'color', icon: 'fas fa-font', tooltip: 'Color' }
    ];
    var ALIGN_ITEMS = [
        { type: 'justifyLeft', icon: 'fas fa-align-left', tooltip: 'Align Left' },
        { type: 'justifyCenter', icon: 'fas fa-align-center', tooltip: 'Align Center' },
        { type: 'justifyRight', icon: 'fas fa-align-right', tooltip: 'Align Right' },
        { type: 'justifyFull', icon: 'fas fa-align-justify', tooltip: 'Align Justify' },
    ];
    var IMAGE_HREF_ITEMS = [
        { type: 'createLink', icon: 'fas fa-paperclip', tooltip: 'Insert Link', linkURLPlaceholder: 'Enter Link URL' },
        { type: 'insertImage', icon: 'fas fa-image', tooltip: 'Insert Image', imageURLPlaceholder: 'Enter image URL' },
    ];
    var LIST_ITEMS = [
        { type: 'insertOrderedList', icon: 'fas fa-list-ol', tooltip: 'Numbered List' },
        { type: 'insertUnorderedList', icon: 'fas fa-list-ul', tooltip: 'Ordered List' },
    ];

    var VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        // tslint:disable-next-line: no-use-before-declare
        useExisting: core.forwardRef(function () { return MdbWysiwygComponent; }),
        multi: true
    };
    var MdbWysiwygComponent = /** @class */ (function () {
        function MdbWysiwygComponent(renderer, cdRef, el) {
            this.renderer = renderer;
            this.cdRef = cdRef;
            this.el = el;
            this._value = '';
            this.options = null;
            this.valueChange = new core.EventEmitter();
            this.valueContent = new core.EventEmitter();
            this.isTextStyleDropdownVisible = false;
            this.isTextColorDropdownVisible = false;
            this.selectedTextStyle = 'Paragraph';
            this.selectedTextColor = '#000';
            this.showTooltips = true;
            this.optionsLength = 0;
            this.textStyleItems = TEXT_STYLE_ITEMS;
            this.textColorItems = TEXT_COLOR_ITEMS;
            this.textDecorationItems = TEXT_DECORATION_ITEMS;
            this.alignItems = ALIGN_ITEMS;
            this.imageHrefItems = IMAGE_HREF_ITEMS;
            this.listItems = LIST_ITEMS;
            this._destroy = new rxjs.Subject();
            this._htmlIsVisible = false;
            this.wysiwygValueChange$ = new rxjs.Subject();
            this.onChange = function (_) { };
            this.onTouched = function () { };
        }
        Object.defineProperty(MdbWysiwygComponent.prototype, "value", {
            get: function () {
                return this._value;
            },
            set: function (newValue) {
                if (newValue || newValue === '') {
                    this._value = newValue;
                    this.textarea.nativeElement.innerHTML = newValue;
                }
                else {
                    this._value = '';
                    this.textarea.nativeElement.innerHTML = '';
                }
                this.valueChange.emit(newValue);
                this.wysiwygValueChange$.next(newValue);
            },
            enumerable: false,
            configurable: true
        });
        MdbWysiwygComponent.prototype.valueChange$ = function () {
            return this.wysiwygValueChange$;
        };
        MdbWysiwygComponent.prototype.writeValue = function (value) {
            if (value || value === '') {
                this.value = value;
                this.textarea.nativeElement.innerHTML = this.value;
                this.wysiwygValueChange$.next(value);
                this.valueChange.emit(value);
                this.onChange(value);
                this.onTouched();
            }
        };
        MdbWysiwygComponent.prototype.valueChanged = function () {
            this.onChange(this.value);
        };
        MdbWysiwygComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        MdbWysiwygComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        MdbWysiwygComponent.prototype.setDisableState = function (isDisabled) { };
        MdbWysiwygComponent.prototype.selectTextStyle = function (item, event) {
            event.preventDefault();
            if (!this.options || !this.options.translations) {
                this.selectedTextStyle =
                    item.text === 'Heading'
                        ? item.text + item.selector.toString().replace('h', ' ')
                        : item.text;
            }
            else {
                this.selectedTextStyle =
                    item.text === this.options.translations['textElements'].heading
                        ? item.text + item.selector.toString().replace('h', ' ')
                        : item.text;
            }
            this.isTextStyleDropdownVisible = false;
            if (this.textStyleItems.findIndex(function (el) { return el.text === window.getSelection().anchorNode.textContent; }) === -1) {
                document.execCommand('formatBlock', false, "<" + item.selector + ">");
            }
        };
        MdbWysiwygComponent.prototype.selectTextColor = function (item, event) {
            event.preventDefault();
            this.selectedTextColor = item.color;
            this.isTextColorDropdownVisible = false;
            document.execCommand('styleWithCSS', false, item.color);
            document.execCommand('foreColor', false, item.color);
        };
        MdbWysiwygComponent.prototype.performAction = function (action, event) {
            event.preventDefault();
            var selection = window.getSelection().anchorNode &&
                window.getSelection().anchorNode.parentElement.firstChild;
            if (selection) {
                if (this.textStyleItems.findIndex(function (el) { return el.text === window.getSelection().anchorNode.textContent; }) === -1) {
                    if (action === 'color') {
                        this.isTextColorDropdownVisible = !this.isTextColorDropdownVisible;
                    }
                    document.execCommand(action);
                }
                var actions = [
                    'justifyLeft',
                    'justifyCenter',
                    'justifyRight',
                    'justifyFull',
                    'createLink',
                    'insertImage'
                ];
                if (actions.find(function (el) { return el === action; })) {
                    var value = null;
                    if (action === 'createLink' || action === 'insertImage') {
                        var link = window.getSelection().getRangeAt(0).toString() || 'https://';
                        if (!this.options || !this.options.translations) {
                            value = prompt('Insert URL: ', link);
                        }
                        else {
                            value = prompt("" + (action === 'createLink'
                                ? this.options.translations['imageAndLink'].linkURLPlaceholder
                                : this.options.translations['imageAndLink'].imageURLPlaceholder), link);
                        }
                        document.execCommand('delete', false);
                    }
                    document.execCommand(action, false, value);
                }
            }
        };
        MdbWysiwygComponent.prototype.toggleHTMLCode = function () {
            if (!this._htmlIsVisible) {
                this.textarea.nativeElement.textContent = this.textarea.nativeElement.innerHTML;
                this._htmlIsVisible = true;
            }
            else {
                this.textarea.nativeElement.innerHTML = this.textarea.nativeElement.innerText;
                this._htmlIsVisible = false;
            }
        };
        MdbWysiwygComponent.prototype.changeTooltipTranslation = function (array, el, value) {
            array.forEach(function (element) {
                element[el] = value[element.type];
            });
        };
        MdbWysiwygComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.value || this.value === '') {
                this.textarea.nativeElement.innerHTML = this.value;
            }
            else {
                this.textarea.nativeElement.innerHTML = '';
            }
            this.renderer.listen(this.textarea.nativeElement, 'click', function () {
                _this.isTextStyleDropdownVisible = false;
                _this.isTextColorDropdownVisible = false;
            });
            rxjs.fromEvent(this.el.nativeElement, 'input')
                .pipe(operators.debounceTime(100), operators.distinctUntilChanged(), operators.takeUntil(this._destroy))
                .subscribe(function (event) {
                _this.wysiwygValueChange$.next(event.target.innerHTML);
                _this.valueChange.emit(event.target.innerHTML);
                _this.onChange(event.target.innerHTML);
                _this.valueContent.emit(event.target.textContent);
                _this.onTouched();
            });
            if (this.options) {
                this.optionsLength = Object.entries(this.options).length;
                if (this.options.translations) {
                    Object.entries(this.options.translations).forEach(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        switch (key) {
                            case 'textElements': {
                                _this.selectedTextStyle = value.paragraph;
                                _this.changeTooltipTranslation(TEXT_STYLE_ITEMS, 'text', value);
                                break;
                            }
                            case 'textStyle': {
                                _this.changeTooltipTranslation(TEXT_DECORATION_ITEMS, 'tooltip', value);
                                break;
                            }
                            case 'textAlign': {
                                _this.changeTooltipTranslation(ALIGN_ITEMS, 'tooltip', value);
                                break;
                            }
                            case 'imageAndLink': {
                                _this.changeTooltipTranslation(IMAGE_HREF_ITEMS, 'tooltip', value);
                                break;
                            }
                            case 'lists': {
                                _this.changeTooltipTranslation(LIST_ITEMS, 'tooltip', value);
                                break;
                            }
                            default: {
                                _this.options.translations['showHTML'] = value;
                                break;
                            }
                        }
                        if (!_this.options.translations['showHTML']) {
                            _this.options.translations['showHTML'] = 'Show HTML Code';
                        }
                    });
                }
                if (this.options.colors) {
                    this.textColorItems = [];
                    Object.entries(this.options.colors).forEach(function (color) {
                        _this.textColorItems.push({ name: color[0], color: color[1] });
                    });
                }
            }
            if (!this.options || Object.entries(this.options).length === 0) {
                this.showTooltips = true;
            }
            else if (this.options && !this.options.tooltips) {
                this.showTooltips = false;
            }
            this.cdRef.detectChanges();
        };
        MdbWysiwygComponent.prototype.ngOnDestroy = function () {
            this._destroy.next();
            this._destroy.complete();
        };
        return MdbWysiwygComponent;
    }());
    MdbWysiwygComponent.decorators = [
        { type: core.Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: 'mdb-wysiwyg',
                    template: "<div class=\"\">\n  <textarea id=\"demo\" style=\"width: 0px; height: 0px; visibility: hidden\" cols=\"30\" rows=\"10\"></textarea>\n  <div class=\"mdb-wysiwyg-container\">\n    <div class=\"mdb-wysiwyg-toolbar\">\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <div class=\"mdb-wysiwyg-toolbar-dropdown\">\n          <div class=\"mdb-wysiwyg-toolbar-dropdown-toggle text-style-dropdown\"\n               (click)=\"isTextStyleDropdownVisible = !isTextStyleDropdownVisible\">\n            <button type=\"button\" class=\"btn-clear\">{{selectedTextStyle}}</button>\n          </div>\n          <ul class=\"mdb-wysiwyg-toolbar-options-list\"\n              [ngClass]=\"{'visible': isTextStyleDropdownVisible, 'hidden': !isTextStyleDropdownVisible}\">\n            <li *ngFor=\"let item of textStyleItems; let i = index\" (click)=\"selectTextStyle(item, $event)\">\n              <button type=\"button\" class=\"btn-clear w-100\">{{item.type === 'heading' ? item.text + ' ' + i : item.text}}</button>\n            </li>\n          </ul>\n        </div>\n      </div>\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <button type=\"button\" *ngFor=\"let item of textDecorationItems; let i = index\"\n                (click)=\"performAction(item.type, $event)\"\n                [ngClass]=\"{'btn-clear': true, 'mdb-wysiwyg-toolbar-dropdown-toggle': i == textDecorationItems.length - 1}\"\n                [mdbTooltip]=\"showTooltips ? item.tooltip : ''\" placement=\"bottom\" triggers=\"hover click\">\n          <i class=\"{{item.icon}}\"></i>\n        </button>\n\n        <ul class=\"mdb-wysiwyg-toolbar-options-list mdb-wysiwyg-toolbar-color-palette\"\n            [ngClass]=\"{'visible': isTextColorDropdownVisible, 'hidden': !isTextColorDropdownVisible}\">\n          <li *ngFor=\"let item of textColorItems; let i = index\" (click)=\"selectTextColor(item, $event)\"\n              [ngClass]=\"{'visible': isTextColorDropdownVisible, 'hidden': !isTextColorDropdownVisible}\">\n            <button type=\"button\" class=\"btn-clear\" [ngStyle]=\"{'background-color': item.color}\"></button>\n          </li>\n        </ul>\n      </div>\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <ul class=\"mdb-wysiwyg-toolbar-options\">\n          <li *ngFor=\"let item of alignItems; let i = index\" (click)=\"performAction(item.type, $event)\">\n            <button type=\"button\" class=\"btn-clear\" [mdbTooltip]=\"showTooltips  ? item.tooltip : ''\" placement=\"bottom\"\n                    triggers=\"hover click\">\n              <i class=\"{{item.icon}}\"></i>\n            </button>\n          </li>\n        </ul>\n      </div>\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <ul class=\"mdb-wysiwyg-toolbar-options\">\n          <li *ngFor=\"let item of imageHrefItems; let i = index\" (click)=\"performAction(item.type, $event)\">\n            <button type=\"button\" class=\"btn-clear\" [mdbTooltip]=\"showTooltips  ? item.tooltip : ''\" placement=\"bottom\"\n                    triggers=\"hover click\">\n              <i class=\"{{item.icon}}\"></i>\n            </button>\n          </li>\n        </ul>\n      </div>\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <ul class=\"mdb-wysiwyg-toolbar-options\">\n          <li *ngFor=\"let item of listItems; let i = index\" (click)=\"performAction(item.type, $event)\">\n            <button type=\"button\" class=\"btn-clear\" [mdbTooltip]=\"showTooltips  ? item.tooltip : ''\" placement=\"bottom\"\n                    triggers=\"hover click\">\n              <i class=\"{{item.icon}}\"></i>\n            </button>\n          </li>\n        </ul>\n      </div>\n      <div class=\"mdb-wysiwyg-toolbar-group\">\n        <ul class=\"mdb-wysiwyg-toolbar-options\">\n          <li (click)=\"toggleHTMLCode()\">\n            <button type=\"button\" class=\"btn-clear\" *ngIf=\"options && options.tooltips\"\n                    [mdbTooltip]=\"options && options.translations ? options.translations['showHTML'] : 'Show HTML Code'\"\n                    placement=\"bottom\" triggers=\"hover click\">\n              <i class=\"fas fa-code\"></i>\n            </button>\n\n            <button type=\"button\" class=\"btn-clear\" *ngIf=\"options && !options.tooltips; else showHTML\"\n                    [mdbTooltip]=\"options && !options.tooltips && showTooltips ? 'Show HTML Code' : ''\"\n                    placement=\"bottom\"\n                    triggers=\"hover click\">\n              <i class=\"fas fa-code\"></i>\n            </button>\n\n            <ng-template #showHTML>\n              <button type=\"button\" class=\"btn-clear\" *ngIf=\"optionsLength === 0\"\n                      [mdbTooltip]=\"optionsLength === 0 ? 'Show HTML Code' : ''\" placement=\"bottom\" triggers=\"hover click\">\n                <i class=\"fas fa-code\"></i>\n              </button>\n            </ng-template>\n          </li>\n        </ul>\n      </div>\n    </div>\n    <div class=\"mdb-wysiwyg-textarea\" #textarea contenteditable=\"true\">\n    </div>\n  </div>\n</div>\n",
                    providers: [VALUE_ACCESSOR],
                    encapsulation: core.ViewEncapsulation.None,
                    styles: [".mdb-wysiwyg-container{display:block;border:1px solid #ededed;border-radius:3px 3px 0 0}.mdb-wysiwyg-toolbar-dropdown-toggle.text-style-dropdown:after{content:\"\";display:inline-block;border-style:solid;border-width:3px;border-color:#888 transparent transparent;right:0;margin-left:0}.mdb-wysiwyg-toolbar-color-palette{margin-top:26px}.btn-clear{background:transparent;border:0;cursor:pointer}.mdb-wysiwyg-toolbar{display:flex;flex-wrap:wrap;justify-content:space-between;padding-left:8px;padding-right:8px;border-top:1px solid #fff;border-bottom:1px solid #d9dad9;border-radius:3px 3px 0 0;background:#f4f4f4;background:linear-gradient(#f4f4f4,#efeeee)}.mdb-wysiwyg-toolbar-options{list-style:none;margin:5px 0;padding-left:0}.mdb-wysiwyg-toolbar-options li{display:inline-block;padding-left:0}.mdb-wysiwyg-toolbar-options button{color:#555;background:transparent;border:none;border-radius:3px;cursor:pointer}.mdb-wysiwyg-toolbar-options button:focus{outline:0}.mdb-wysiwyg-toolbar-options button.active{border:1px solid #ccc;box-shadow:inset 0 0 13px #00000025}.mdb-wysiwyg-toolbar-options-list{visibility:hidden;position:absolute;background:#fff;box-shadow:0 2px 4px #0003;border:1px solid rgba(0,0,0,.2);list-style:none;font-size:.8rem;padding-left:0;z-index:99;opacity:0;transition:opacity .2s}.mdb-wysiwyg-toolbar-options-list.visible{visibility:visible;opacity:1}.mdb-wysiwyg-toolbar-options-list li{display:block}.mdb-wysiwyg-toolbar-options-list button{display:block;padding:4px 10px;color:#000;cursor:pointer}.mdb-wysiwyg-toolbar-options-list button:hover{background:#eee}.mdb-wysiwyg-toolbar-group{display:flex}.mdb-wysiwyg-toolbar-dropdown{align-self:center}.mdb-wysiwyg-toolbar-dropdown-toggle{font-size:.875rem;color:#555;border:0;background:transparent;cursor:pointer}.mdb-wysiwyg-toolbar-dropdown-toggle:focus{outline:0}.mdb-wysiwyg-toolbar-dropdown-toggle:after{content:\"\";display:inline-block;border-style:solid;border-width:3px;border-color:#888 transparent transparent;right:0;margin-left:5px}.mdb-wysiwyg-toolbar-color-palette{display:grid;grid-template-columns:repeat(5,1fr);grid-gap:3px;padding:3px}.mdb-wysiwyg-toolbar-color-palette li{padding:0}.mdb-wysiwyg-toolbar-color-palette button{display:block;width:14px;height:14px;border:1px solid rgba(0,0,0,.15);border-radius:0}.mdb-wysiwyg-textarea{display:block;padding:15px;width:100%;height:450px;border:0;resize:none;overflow:auto;background:#fff}.mdb-wysiwyg-textarea.show-raw{white-space:pre-line}.mdb-wysiwyg-textarea:focus{outline:0}.mdb-wysiwyg-textarea b,.mdb-wysiwyg-textarea strong{font-weight:bold}\n"]
                },] }
    ];
    MdbWysiwygComponent.ctorParameters = function () { return [
        { type: core.Renderer2 },
        { type: core.ChangeDetectorRef },
        { type: core.ElementRef }
    ]; };
    MdbWysiwygComponent.propDecorators = {
        value: [{ type: core.Input }],
        options: [{ type: core.Input }],
        valueChange: [{ type: core.Output }],
        valueContent: [{ type: core.Output }],
        textarea: [{ type: core.ViewChild, args: ['textarea', { static: true },] }]
    };

    var MdbWysiwygModule = /** @class */ (function () {
        function MdbWysiwygModule() {
        }
        return MdbWysiwygModule;
    }());
    MdbWysiwygModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [MdbWysiwygComponent],
                    imports: [
                        common.CommonModule,
                        angularBootstrapMd.TooltipModule.forRoot()
                    ],
                    exports: [MdbWysiwygComponent, common.CommonModule]
                },] }
    ];

    /*
     * Public API Surface of mdb-wysiwyg
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ALIGN_ITEMS = ALIGN_ITEMS;
    exports.IMAGE_HREF_ITEMS = IMAGE_HREF_ITEMS;
    exports.LIST_ITEMS = LIST_ITEMS;
    exports.MdbWysiwygComponent = MdbWysiwygComponent;
    exports.MdbWysiwygModule = MdbWysiwygModule;
    exports.TEXT_COLOR_ITEMS = TEXT_COLOR_ITEMS;
    exports.TEXT_DECORATION_ITEMS = TEXT_DECORATION_ITEMS;
    exports.TEXT_STYLE_ITEMS = TEXT_STYLE_ITEMS;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
