// mdb-color-picker/esm2015/lib/directives/mdb-color-draggable.directive.js
import { Directive, ElementRef, HostBinding, HostListener } from '@angular/core';
export class MdbColorDraggableDirective {
    constructor(el) {
        this.el = el;
        this.elementPosition = { x: 0 };
        this.movable = true;
    }
    onStartMove(event) {
        this.isElementMoving = true;
        this.startMoving(event);
    }
    onMove(event) {
        if (!this.isElementMoving) {
            return;
        }
        this.movingOver(event);
    }
    onEndMove() {
        if (!this.isElementMoving) {
            return;
        }
        this.isElementMoving = false;
    }
    startMoving(event) {
        if (event.type === 'mousedown') {
            this.startingPosition = {
                x: event.clientX - this.elementPosition.x
            };
        }
        else {
            this.startingPosition = {
                x: event.changedTouches[0].clientX - this.elementPosition.x
            };
        }
    }
    movingOver(event) {
        if (event.type === 'mousemove') {
            this.elementPosition.x = event.clientX - this.startingPosition.x;
        }
        else {
            this.elementPosition.x =
                event.changedTouches[0].clientX - this.startingPosition.x;
        }
        this.updatePosition();
    }
    moveTo(value) {
        this.elementPosition.x = value;
        this.updatePosition();
    }
    updatePosition() {
        if (this.elementPosition.x >= 1 &&
            this.elementPosition.x <= this.el.nativeElement.parentElement.offsetWidth) {
            requestAnimationFrame(() => {
                this.el.nativeElement.style.transform = `
                translateX(${this.elementPosition.x}px)
              `;
            });
        }
    }
}
MdbColorDraggableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mdbColorDraggable]'
            },] }
];
MdbColorDraggableDirective.ctorParameters = () => [
    { type: ElementRef }
];
MdbColorDraggableDirective.propDecorators = {
    movable: [{ type: HostBinding, args: ['class.mdb-movable',] }],
    onStartMove: [{ type: HostListener, args: ['mousedown', ['$event'],] }, { type: HostListener, args: ['touchstart', ['$event'],] }],
    onMove: [{ type: HostListener, args: ['document:mousemove', ['$event'],] }, { type: HostListener, args: ['touchmove', ['$event'],] }],
    onEndMove: [{ type: HostListener, args: ['document:mouseup',] }, { type: HostListener, args: ['touchend',] }]
};
