// mdb-color-picker/esm2015/lib/directives/mdb-color-picker-toggle.directive.js
import { Directive, HostListener, Input, Renderer2 } from '@angular/core';
export class MdbColorPickerToggleDirective {
    constructor(renderer) {
        this.renderer = renderer;
        this.transitionTime = 200;
        this._isPickerVisible = false;
        this._clickEventTarget = null;
    }
    onClick(event) {
        this._clickEventTarget = event.target;
        this.toggleColorPicker();
    }
    onWindowClick(event) {
        if (event.target !== this._clickEventTarget &&
            this._isPickerVisible &&
            !this._getClosestEl(event.target, 'mdb-color-picker')) {
            this.toggleColorPicker();
        }
    }
    toggleColorPicker() {
        const opacityStyle = this._isPickerVisible ? '0' : '1';
        const visibilityStyle = this._isPickerVisible ? 'hidden' : 'visible';
        this.renderer.setStyle(this.mdbColorPickerToggle.el.nativeElement, 'transition', `opacity ${this.transitionTime}ms`);
        this.renderer.setStyle(this.mdbColorPickerToggle.el.nativeElement, 'opacity', opacityStyle);
        this.renderer.setStyle(this.mdbColorPickerToggle.el.nativeElement, 'visibility', 'visible');
        setTimeout(() => {
            this.renderer.setStyle(this.mdbColorPickerToggle.el.nativeElement, 'visibility', visibilityStyle);
        }, this.transitionTime);
        this._isPickerVisible = !this._isPickerVisible;
    }
    _getClosestEl(el, selector) {
        for (; el && el !== document; el = el.parentNode) {
            if (el.matches && el.matches(selector)) {
                return el;
            }
        }
        return null;
    }
}
MdbColorPickerToggleDirective.decorators = [
    { type: Directive, args: [{ selector: '[mdbColorPickerToggle]' },] }
];
MdbColorPickerToggleDirective.ctorParameters = () => [
    { type: Renderer2 }
];
MdbColorPickerToggleDirective.propDecorators = {
    mdbColorPickerToggle: [{ type: Input }],
    transitionTime: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
    onWindowClick: [{ type: HostListener, args: ['window:click', ['$event'],] }]
};
