// mdb-filter/esm2015/lib/mdb-filter.service.js
import { Injectable } from '@angular/core';
import { Subject } from "rxjs";
import * as i0 from "@angular/core";
export class MdbFilterService {
    constructor() {
        this._items = [];
        this._items$ = new Subject();
        this._filter = '';
        this._filter$ = new Subject();
        this._previousCurrentFilter = { previousValue: '', currentValue: '' };
        this._previousCurrentFilter$ = new Subject();
    }
    setPreviousCurrentFilter(value) {
        this._previousCurrentFilter = value;
        this._previousCurrentFilter$.next(this.getPreviousCurrentFilter());
    }
    getPreviousCurrentFilter() {
        return this._previousCurrentFilter;
    }
    previousCurrentFilterChange() {
        return this._previousCurrentFilter$;
    }
    setFilter(value) {
        this._filter = value;
        this._filter$.next(this.getFilter());
    }
    getFilter() {
        return this._filter;
    }
    filterChange() {
        return this._filter$;
    }
    setItems(items) {
        this._items = items;
        this._items$.next(this.getItems());
    }
    getItems() {
        return this._items;
    }
    itemsChange() {
        return this._items$;
    }
}
MdbFilterService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MdbFilterService_Factory() { return new MdbFilterService(); }, token: MdbFilterService, providedIn: "root" });
MdbFilterService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
MdbFilterService.ctorParameters = () => [];
