// mdb-wysiwyg/esm2015/lib/mdb-defaults.js
export const TEXT_STYLE_ITEMS = [
    { type: 'paragraph', selector: 'p', text: 'Paragraph' },
    { type: 'heading', selector: 'h1', text: 'Heading' },
    { type: 'heading', selector: 'h2', text: 'Heading' },
    { type: 'heading', selector: 'h3', text: 'Heading' },
    { type: 'heading', selector: 'h4', text: 'Heading' },
    { type: 'heading', selector: 'h5', text: 'Heading' },
    { type: 'heading', selector: 'h6', text: 'Heading' },
    { type: 'preformatted', selector: 'pre', text: 'Preformatted' },
];
export const TEXT_COLOR_ITEMS = [
    { name: 'red', color: '#f44336' },
    { name: 'pink', color: '#e91e63' },
    { name: 'purple', color: '#9c27b0' },
    { name: 'deep-purple', color: '#673ab7' },
    { name: 'indigo', color: '#3f51b5' },
    { name: 'blue', color: '#2196f3' },
    { name: 'light-blue', color: '#03a9f4' },
    { name: 'cyan', color: '#00bcd4' },
    { name: 'teal', color: '#009688' },
    { name: 'green', color: '#4caf50' },
    { name: 'light-green', color: '#8bc34a' },
    { name: 'lime', color: '#cddc39' },
    { name: 'yellow', color: '#ffeb3b' },
    { name: 'amber', color: '#ffc107' },
    { name: 'orange', color: '#ff9800' },
    { name: 'deep-orange', color: '#ff5722' },
    { name: 'brown', color: '#795548' },
    { name: 'grey', color: '#9e9e9e' },
    { name: 'blue-grey', color: '#607d8b' },
    { name: 'mdb-color', color: '#45526e' },
    { name: 'white', color: '#fff' },
    { name: 'black', color: '#000' },
];
export const TEXT_DECORATION_ITEMS = [
    { type: 'bold', icon: 'fas fa-bold', tooltip: 'Bold' },
    { type: 'italic', icon: 'fas fa-italic', tooltip: 'Italic' },
    { type: 'strikethrough', icon: 'fas fa-strikethrough', tooltip: 'Strikethrough' },
    { type: 'underline', icon: 'fas fa-underline', tooltip: 'Underline' },
    { type: 'color', icon: 'fas fa-font', tooltip: 'Color' }
];
export const ALIGN_ITEMS = [
    { type: 'justifyLeft', icon: 'fas fa-align-left', tooltip: 'Align Left' },
    { type: 'justifyCenter', icon: 'fas fa-align-center', tooltip: 'Align Center' },
    { type: 'justifyRight', icon: 'fas fa-align-right', tooltip: 'Align Right' },
    { type: 'justifyFull', icon: 'fas fa-align-justify', tooltip: 'Align Justify' },
];
export const IMAGE_HREF_ITEMS = [
    { type: 'createLink', icon: 'fas fa-paperclip', tooltip: 'Insert Link', linkURLPlaceholder: 'Enter Link URL' },
    { type: 'insertImage', icon: 'fas fa-image', tooltip: 'Insert Image', imageURLPlaceholder: 'Enter image URL' },
];
export const LIST_ITEMS = [
    { type: 'insertOrderedList', icon: 'fas fa-list-ol', tooltip: 'Numbered List' },
    { type: 'insertUnorderedList', icon: 'fas fa-list-ul', tooltip: 'Ordered List' },
];
