// ng-uikit-pro-standard/esm2015/lib/pro/timepicker/timepicker.directive.js
import { Directive, Input, forwardRef, ElementRef, EventEmitter, HostListener, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export const MDB_TIMEPICKER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    // tslint:disable-next-line: no-use-before-declare
    useExisting: forwardRef(() => MdbTimePickerDirective),
    multi: true,
};
export class MdbTimePickerDirective {
    constructor(el) {
        this.el = el;
        this._valueChange = new EventEmitter();
        this.onChange = () => { };
        this.onTouched = () => { };
    }
    set value(value) {
        this._value = value;
        this._valueChange.emit(this._value);
        if (value) {
            this.el.nativeElement.value = value;
        }
        else {
            this.el.nativeElement.value = null;
        }
    }
    get value() {
        return this._value;
    }
    handleInput(event) {
        this.onChange(event.target.value);
        this._valueChange.emit(event.target.value);
    }
    ngOnInit() {
        this.mdbTimePicker.setInput(this);
        this._valueChange.emit(this._value);
        this.mdbTimePicker._selectionChange$.subscribe(selectedValue => {
            this.value = selectedValue;
            this.onChange(selectedValue);
        });
    }
    writeValue(value) {
        this.value = value;
        this.mdbTimePicker._selectionChange$.next(this._value);
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
}
MdbTimePickerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mdbTimePicker]',
                // tslint:disable-next-line: no-host-metadata-property
                host: { '(blur)': 'onTouched()' },
                providers: [MDB_TIMEPICKER_VALUE_ACCESSOR],
            },] }
];
MdbTimePickerDirective.ctorParameters = () => [
    { type: ElementRef }
];
MdbTimePickerDirective.propDecorators = {
    mdbTimePicker: [{ type: Input }],
    value: [{ type: Input }],
    handleInput: [{ type: HostListener, args: ['input', ['$event'],] }]
};
