// mdb-filter/esm2015/lib/mdb-filter-content/mdb-filter-content.component.js
import { Component, ContentChildren, ElementRef, QueryList } from '@angular/core';
import { MdbFilterService } from "../mdb-filter.service";
import { MdbFilterItemComponent } from "../mdb-filter-item/mdb-filter-item.component";
export class MdbFilterContentComponent {
    constructor(filterService) {
        this.filterService = filterService;
        this.filterItems = new QueryList();
    }
    ngAfterContentInit() {
        this.filterService.setItems(this.filterItems.map(el => el.nativeElement));
    }
}
MdbFilterContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'mdb-filter-content',
                template: "<div class=\"filter-gallery\">\n  <ng-content></ng-content>\n</div>\n",
                styles: [""]
            },] }
];
MdbFilterContentComponent.ctorParameters = () => [
    { type: MdbFilterService }
];
MdbFilterContentComponent.propDecorators = {
    filterItems: [{ type: ContentChildren, args: [MdbFilterItemComponent, { read: ElementRef, descendants: true },] }]
};
