// mdb-filter/esm2015/lib/mdb-filter/mdb-filter.component.js
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { MdbFilterService } from "../mdb-filter.service";
export class MdbFilterComponent {
    constructor(filterService) {
        this.filterService = filterService;
        this._filter = '';
        this.filterService.filterChange().subscribe((filter) => {
            this.filter = filter;
        });
    }
    set filter(value) {
        const previousValue = this.filter;
        this._filter = value;
        this.filterService.setPreviousCurrentFilter({ previousValue: previousValue, currentValue: this._filter });
    }
    get filter() {
        return this._filter;
    }
}
MdbFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'mdb-filter',
                template: "<div class=\"filter filter-basic\">\n<ng-content></ng-content>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".filter .filter-gallery{position:relative}.filter .item{display:block;z-index:1}.filter .item-panoramic{width:410px}.filter .item-content{position:relative;width:100%;height:100%}.filter .item-content img{width:100%;height:auto}.filter-draggable .item{cursor:move}.filter-fillgaps .item{cursor:move}\n"]
            },] }
];
MdbFilterComponent.ctorParameters = () => [
    { type: MdbFilterService }
];
MdbFilterComponent.propDecorators = {
    filter: [{ type: Input, args: ['filter',] }]
};
